draft-ietf-ppm-dap.md

```diff
@@ -1264,30 +1265,37 @@ enum {
 } PrepareStepState;
 
 struct {
-  ReportID report_id;
   PrepareStepState prepare_step_state;
   select (PrepareStep.prepare_step_state) {
-    case continued: opaque prep_msg<0..2^32-1>; /* VDAF preparation message */
+    case continued: opaque prep_msg<0..2^32-1>; /* VDAF preparation message or share */
     case finished:  Empty;
     case failed:    ReportShareError;
   };
 } PrepareStep;
 
 struct {
-  PrepareStep prepare_steps<1..2^32-1>;
-} AggregationJobResp;
+  ReportID report_id;
+  PrepareStep init_prep_step;
+  PrepareStep stepped_prep_step;
```

Comment:
Hmm, what's going on here? IIUC this should carry the first-round prep message. Why do we have two PrepareStep's

Reply:
I agree, there should just be the current round combined/broadcast prepare message in this message. Let's look at what the Helper does during initialization, and let's imagine this is a VDAF with more than one round.

~~~
prep_msg = VDAF.prep_shares_to_prep(agg_param, [leader_prep_share, helper_prep_share])
(next_prep_state, next_prep_share) = VDAF.prep_next(prep_state, prep_msg)
~~~

`leader_prep_share` is the leader's share of the 1st round prepare message, and is what was transmitted in the `AggregationJobInitReq`. `helper_prep_share` is the helper's share of the 1st round prepare message was derived from the encrypted report share by the helper. `prep_msg` is the combined/broadcast 1st round prepare message.

Helper then uses `prep_msg` and `prep_state`, which is its round 0 state, to advance to round 1. The outputs are the helper's round 1 state and the helper's share of round 2's prepare message. In order to advance the leader to round 1, the helper needs to send the leader `prep_msg`, so that the leader can then use its round 0 state and the round 1 combined/broadcast prepare message to advance to round 1. The leader doesn't need `next_prep_share` because when it continues, it's going to send its round 2 prepare message share to the helper in an AggregationJobContinueReq.

However, this discussion suggests a further possible optimization: if the helper did send the round 2 prepare message share along with the combined round 1 prepare message, then the leader could:

- advance to round 1 using the round 1 prepare message, obtaining its share of the round 2 prepare message
- combine its round 2 prepare message with the helper's share of the round 2 prepare message, yielding the round 2 combined prepare message
- advance itself to round 2
- transmit the round 2 combined prepare message to the helper in an `AggregationJobContinueReq`

Basically, the leader and helper would take turns combining prepare message shares and instructing the other aggregator to advance its round, which could eliminate round trips for VDAFs with more than one round.

Reply:
This change makes the first prepare message be computed from the prepare message shares by the Helper, and keeps all following prepare messages computed by the Leader. I went this way because it keeps more of the work being done by the Leader, and it's the "minimal" change to behavior in the sense that it changes the least about the protocol.

Tim, your first suggestion is equivalent to what is written in this PR except that the Helper is responsible for computing all of the prepare messages, because it changes things to have the leader send a prepare message share rather than a prepare message with each continuation. Your second suggestion is what is written down in this PR -- except that the Leader/Helper doesn't "take turns" (I'm not sure there's an advantage to implementing things this way). Both suggestions are the same number of round-trips, equal to the VDAF's `ROUNDS` parameter, and thus optimal in number of round-trips.

Reply:
(I guess the question is: who do we want to compute the prepare message, i.e. call `prep_shares_to_prep`, after initialization? Using this trick requires the Helper to call `prep_shares_to_prep` during initialization, but after that we can spell the protocol out to allow either the Leader or the Helper call `prep_shares_to_prep` on each continuation round. As-written, the Leader makes these calls; we could change things to have the Helper make these calls instead, or something more advanced like a turn-taking strategy. IMO, I'm happy with either the Leader or the Helper taking on this role, but don't think more advanced strategies where the Leader & Helper take turns are worth the protocol complexity.)

Reply:
@tgeoghegan I don't think the change you suggest actually reduces the number of HTTP requests, does it? I don't see the benefit, and I think it makes things more complex.

@branlwyd  re what is the minimal change: I think this PR should make the following change: Instead of having the Leader run prep_shares_to_prep(), have the Helper do it. That would mean the Leader's request contains its perp share, and the Helper's response contains the prep message (i.e., the broadcast message).

Reply:
Hopefully this will be more clear after #422.

Reply:
We went with an approach that implements Tim's ping-pong idea, reducing the number of network round trips by about half.